### ISSUE TITLE
Include ECL/EQL/Quicklisp in the source tree?

### ISSUE THREAD

**Issue by deadtrickster** [nEXT-Browser/nEXT#7]
Maybe as a git submodules. The idea is to have reproducible/self-contained builds.

**Comment by jmercouris** [comment]
I think this is a really good idea, there's also a few ways I think we can extend this. Because we want EQL for example to only have webkit enabled during build, maybe we can make a fork of EQL with the eql5.pro file only having webkit enabled, and occasionally we can pull from it or something.

OR

We could do the purely submodule option and have something like nEXT-Browser/Build repository with all of these submodules.

The reason I suggested the first option with the forks is because we have to do some modifications to ECL and EQL, and particularly the case with ECL (I don't know when the next release for Mac OS will come out, which includes some patches for our use case)

**Comment by jmercouris** [comment]
It also occurred to me that maybe we can encapsulate all of this information in different package managers. So what if we had something like "sudo apt-get install next-devel" and it includes all of the frozen versions, and sets up a reproducible environment on ubuntu for example


So in that case we'd have different "devel" installations for different OS' (notably Mac OS and Linux) in a separate repository.

What do you think?

**Comment by deadtrickster** [comment]
I would like to have it without packages, just repository. I remember installing all needed qt packages was a bit headache, like it was compile, fail, install and recompile. I'm thinking about two parts

1) OS(package manager) dependent scripts for installing Qt, build tools, etc (i.e. apt install, pkg install, etc)
2) Locked EQL/ECL (submodule/fork is an option here)

Workflow like this - I pushed Quri stuff and now I want to return and contribute more, all I want is to

```
# git pull
# (optional) init submodules
# ./install-deps
# make
```

I believe as any complex project nEXT will uncover bugs/improvements in the deps (likely EQL/ECL), so it should be easy to contribute back. I think forks + submodules will work.

NOTE: by having EQL/ECL as submodules or in tree I also mean they won't be globally installed, yet build configure properly so it uses this local installation.

**Comment by deadtrickster** [comment]
And regarding my Quri problem - in fact I did `git pull` but for some reason binary still has old code. I guess this is related to this "developer experience" topic.

**Comment by jmercouris** [comment]
@deadtrickster 

Yeah, it's definitely a lot of work to get a functioning developer setup, believe me, I know :D

1. OS dependent scripts: Are you envisioning a set of bash scripts to install the dependencies for each OS?

2. We can definitely fork, it will be a little complex though, seeing as they are on gitlab, I guess we'll have to manually set up two remotes and pull/push from them

Within your proposed workflow, can you please explain what init submodules and install-deps does?

Lastly, are you suggesting setting the install prefix for ECL and EQL to a directory within nEXT?

**Comment by jmercouris** [comment]
Also, I think I know why the make had leftover compile files, only AFTER successful compilation are the temporary build files cleaned from your repository. So what probably happened was

1. You built successfully old version
2. You downloaded new version with new dependencies
3. Run make, make does not finish/work due to old build files
4. New make file therefore does not clean up old build artifacts

**Comment by deadtrickster** [comment]
1. Yes, although I would rephrase it as scripts for the OS default shell.
2. Yes, manual syncing. Unfortunately here we have to reinvent dependency locks.

init submodules - git submodule init & update
install-deps - run script from 1.
Yes, prefix and use that artifacts, not global. In fact it could be possibly to not touch/pollute OS at all, probably event Qt could be configured in local folder. But that I suppose would definitely be harder than just apt install.

**Comment by jmercouris** [comment]
I like it, I think we're definitely onto something. Might I suggest a flag so that the user can choose to install globally or locally? I can make a new repository, not sure what OS you are running (I assume Ubuntu), could you contribute the scripts for the environment setup?

**Comment by deadtrickster** [comment]
Yeah, I'll try to help. Regarding the flag - I don't think it's a good idea. I they want than can install whatever they need. Here we care about nEXT and reproducible builds.

**Comment by jmercouris** [comment]
Alright, so I've made a new repository, and I've invited you to become a member so you can have write access to that repository.

In terms of that, I think you might be right, we can keep the install instructions in the main repository for users who want to install the programs systemwide, and these scripts could be for building.

**Comment by jmercouris** [comment]
I'm closing this issue because we've now created the separate repository.
